### ISSUE TITLE
Crash during application shutdown when win32ui attempts to acquire the GIL after the interpreter is already shut down.

### ISSUE THREAD

**Issue by bob-white** [mhammond/pywin32#1838]
Note that issues in this repository are only for bugs or feature requests in the pywin32.

**If you need support or help using this package, please follow [these instructions](https://github.com/mhammond/pywin32/blob/master/README.md#support)** - support or help requests will be closed without comment.

For all bugs, please provide the following information.

* Expected behavior and actual behavior.
Expected behavior is to not attempt to acquire the GIL after application shutdown has started.

Current behavior is instead the CallWndProc callback in win32uimodule.cpp blindly attempts to acquire the GIL when processing incoming message.

Currently I get the following stacktrace: 
```
  python39_d.dll!PyGILState_Ensure() Line 1356	C
  win32ui_d.pyd!CEnterLeavePython::acquire() Line 653	C++
  win32ui_d.pyd!CEnterLeavePython::CEnterLeavePython() Line 650	C++
  win32ui_d.pyd!Python_check_message(const tagMSG * msg) Line 71	C++
  win32ui_d.pyd!CallWndProc(int nCode, unsigned __int64 wParam, __int64 lParam) Line 99	C++
  user32.dll!fnHkINLPCWPSTRUCTW()	Unknown
```
And inside of PyGilState_Ensure we end up exploding the process at:     `assert(gilstate->autoInterpreterState);`
As at this point, the `autoInterpreterState` is NULL

* Steps to reproduce the problem.
This was discovered with a python interpreter we have embedded in our application, and from what I can infer is probably because we continue to destroy resources _after_ the python interpreter has shutdown, so I'm not certain about general reproducibility.

* Version of Python and pywin32
This is in python 3.9.7 pyin32 build_id is 301

**Comment by mhammond** [comment]
Another solution is to simply never uninitialize Python - who is doing that in your repro?

**Comment by bob-white** [comment]
To answer your question, we're uninitializing python as part of our application shutdown process, and if I re-arrange the order so that the various handles being passed through the hook don't cause a crash, I get a completely separate system complaining that I leaked a bunch of memory, because it relied on the the interpreter shutdown to clean up a bunch of PyObjects. 

#1841 - This however did work for me, basically moving the code that actually acquires the GIL to right before it is needed, instead of pre-emptively grabbing it when we already know that the handle isn't valid.

**Comment by mhammond** [comment]
Fixed by #1841
